File: java/api/analyze/PipelineStage.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.typedb.driver.api.analyze;

import java.util.stream.Stream;

/**
 * Representation of a stage in a <code>Pipeline</code>.
 */
public interface PipelineStage {
    /**
     * @return the pipeline stage variant
     */
    com.typedb.driver.jni.PipelineStageVariant getVariant();

    // Type check methods
    boolean isMatch();

    boolean isInsert();

    boolean isPut();

    boolean isUpdate();

    boolean isDelete();

    boolean isSelect();

    boolean isSort();

    boolean isRequire();

    boolean isOffset();

    boolean isLimit();

    boolean isDistinct();

    boolean isReduce();

    // Conversion methods
    MatchStage asMatch();

    InsertStage asInsert();

    PutStage asPut();

    UpdateStage asUpdate();

    DeleteStage asDelete();

    SelectStage asSelect();

    SortStage asSort();

    RequireStage asRequire();

    OffsetStage asOffset();

    LimitStage asLimit();

    DistinctStage asDistinct();

    ReduceStage asReduce();

    /**
     * Represents a "match" stage: match &lt;block&gt;
     * e.g. {@code match $f isa friendship, links (friend: $x, friend: $y)}
     */
    interface MatchStage extends PipelineStage {
        /**
         * The index into <code>Pipeline.conjunctions</code>
         */
        com.typedb.driver.jni.ConjunctionID block();
    }

    /**
     * Represents an "insert" stage: insert &lt;block&gt;
     * e.g. {@code insert $f isa friendship, links (friend: $x, friend: $y)}
     */
    interface InsertStage extends PipelineStage {
        /**
         * The index into <code>Pipeline.conjunctions</code>
         */
        com.typedb.driver.jni.ConjunctionID block();
    }

    /**
     * Represents a "put" stage: put &lt;block&gt;
     * e.g. {@code put $f isa friendship, links (friend: $x, friend: $y)}
     */
    interface PutStage extends PipelineStage {
        /**
         * The index into <code>Pipeline.conjunctions</code>
         */
        com.typedb.driver.jni.ConjunctionID block();
    }

    /**
     * Represents an "update" stage: update &lt;block&gt;
     * e.g. {@code update $owner has name "John"}
     */
    interface UpdateStage extends PipelineStage {
        /**
         * The index into <code>Pipeline.conjunctions</code>
         */
        com.typedb.driver.jni.ConjunctionID block();
    }

    /**
     * Represents a "delete" stage:
     * <pre>
     * delete
     *     &lt;block&gt;;
     *     &lt;deleted_variables&gt;;
     * </pre>
     * e.g.
     * <pre>
     * delete
     *     has $attribute of $owner; links ($player) of $relation;
     *     $deleted-instance;
     * </pre>
     */
    interface DeleteStage extends PipelineStage {
        /**
         * The index into <code>Pipeline.conjunctions</code>
         */
        com.typedb.driver.jni.ConjunctionID block();

        /**
         * The variables for which the unified concepts are to be deleted.
         */
        Stream<com.typedb.driver.jni.Variable> deletedVariables();
    }

    /**
     * Represents a "select" stage: select &lt;variables&gt;
     * e.g. {@code select $x, $y}
     */
    interface SelectStage extends PipelineStage {
        /**
         * The variables being selected
         */
        Stream<com.typedb.driver.jni.Variable> variables();
    }

    /**
     * Represents a "sort" stage: sort &lt;variables-and-order&gt;
     * e.g. {@code sort $x asc, $y desc}
     */
    interface SortStage extends PipelineStage {
        /**
         * The sort variables and their sort orders
         */
        Stream<? extends SortVariable> variables();

        /**
         * A variable and its sort order
         */
        interface SortVariable {
            /**
             * The variable to sort by
             */
            com.typedb.driver.jni.Variable variable();

            /**
             * The sort order (ascending or descending)
             */
            com.typedb.driver.jni.SortOrder order();
        }
    }

    /**
     * Represents a "require" stage: require &lt;variables&gt;
     * e.g. {@code require $x, $y}
     */
    interface RequireStage extends PipelineStage {
        /**
         * The variables that must be present in the result
         */
        Stream<com.typedb.driver.jni.Variable> variables();
    }

    /**
     * Represents an "offset" stage: offset &lt;offset&gt;
     * e.g. {@code offset 3}
     */
    interface OffsetStage extends PipelineStage {
        /**
         * The number of results to skip
         */
        long offset();
    }

    /**
     * Represents a "limit" stage: limit &lt;limit&gt;
     * e.g. {@code limit 5}
     */
    interface LimitStage extends PipelineStage {
        /**
         * The maximum number of results to return
         */
        long limit();
    }

    /**
     * Represents a "distinct" stage
     */
    interface DistinctStage extends PipelineStage {
    }

    /**
     * Represents a "reduce" stage:
     * <pre>
     * reduce &lt;reducers&gt; groupby &lt;groupby&gt;
     * </pre>
     * e.g.
     * <pre>
     * reduce $sum = sum($v), $count = count groupby $x, $y;
     * </pre>
     */
    interface ReduceStage extends PipelineStage {
        /**
         * The variables to group by
         */
        Stream<com.typedb.driver.jni.Variable> groupBy();

        /**
         * The reducer assignments
         */
        Stream<? extends ReduceAssignment> reducerAssignments();

        /**
         * An assignment of a reducer to a variable
         */
        interface ReduceAssignment {
            /**
             * The variable being assigned to
             */
            com.typedb.driver.jni.Variable assigned();

            /**
             * The reducer being applied
             */
            Reducer reducer();
        }
    }
}

File: python/typedb/api/analyze/pipeline_stage.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

from __future__ import annotations

from abc import ABC, abstractmethod
from typing import TYPE_CHECKING, Iterator

if TYPE_CHECKING:
    from typedb.api.analyze.conjunction_id import ConjunctionID
    from typedb.api.analyze.reducer import Reducer
    from typedb.common.enums import SortOrder
    from typedb.api.analyze.variable import Variable


class PipelineStage(ABC):
    """
    Representation of a stage in a Pipeline.
    """

    @abstractmethod
    def is_match(self) -> bool:
        pass

    @abstractmethod
    def is_insert(self) -> bool:
        pass

    @abstractmethod
    def is_put(self) -> bool:
        pass

    @abstractmethod
    def is_update(self) -> bool:
        pass

    @abstractmethod
    def is_delete(self) -> bool:
        pass

    @abstractmethod
    def is_select(self) -> bool:
        pass

    @abstractmethod
    def is_sort(self) -> bool:
        pass

    @abstractmethod
    def is_require(self) -> bool:
        pass

    @abstractmethod
    def is_offset(self) -> bool:
        pass

    @abstractmethod
    def is_limit(self) -> bool:
        pass

    @abstractmethod
    def is_distinct(self) -> bool:
        pass

    @abstractmethod
    def is_reduce(self) -> bool:
        pass

    @abstractmethod
    def as_match(self) -> "MatchStage":
        pass

    @abstractmethod
    def as_insert(self) -> "InsertStage":
        pass

    @abstractmethod
    def as_put(self) -> "PutStage":
        pass

    @abstractmethod
    def as_update(self) -> "UpdateStage":
        pass

    @abstractmethod
    def as_delete(self) -> "DeleteStage":
        pass

    @abstractmethod
    def as_select(self) -> "SelectStage":
        pass

    @abstractmethod
    def as_sort(self) -> "SortStage":
        pass

    @abstractmethod
    def as_require(self) -> "RequireStage":
        pass

    @abstractmethod
    def as_offset(self) -> "OffsetStage":
        pass

    @abstractmethod
    def as_limit(self) -> "LimitStage":
        pass

    @abstractmethod
    def as_distinct(self) -> "DistinctStage":
        pass

    @abstractmethod
    def as_reduce(self) -> "ReduceStage":
        pass


class MatchStage(PipelineStage, ABC):
    """Represents a 'match' stage: match <block>"""

    @abstractmethod
    def block(self) -> "ConjunctionID":
        """The index into Pipeline.conjunctions."""
        pass


class InsertStage(PipelineStage, ABC):
    """Represents an 'insert' stage: insert <block>"""

    @abstractmethod
    def block(self) -> "ConjunctionID":
        pass


class PutStage(PipelineStage, ABC):
    """Represents a 'put' stage: put <block>"""

    @abstractmethod
    def block(self) -> "ConjunctionID":
        pass


class UpdateStage(PipelineStage, ABC):
    """Represents an 'update' stage: update <block>"""

    @abstractmethod
    def block(self) -> "ConjunctionID":
        pass


class DeleteStage(PipelineStage, ABC):
    """Represents a 'delete' stage."""

    @abstractmethod
    def block(self) -> "ConjunctionID":
        pass

    @abstractmethod
    def deleted_variables(self) -> Iterator["Variable"]:
        """The variables for which the unified concepts are to be deleted."""
        pass


class SelectStage(PipelineStage, ABC):
    """Represents a 'select' stage: select <variables>"""

    @abstractmethod
    def variables(self) -> Iterator["Variable"]:
        pass


class SortStage(PipelineStage, ABC):
    """Represents a 'sort' stage: sort <variables-and-order>"""

    @abstractmethod
    def variables(self) -> Iterator["SortStage.SortVariable"]:
        pass

    class SortVariable(ABC):
        """A variable and its sort order."""

        @abstractmethod
        def variable(self) -> "Variable":
            pass

        @abstractmethod
        def order(self) -> "SortOrder":
            pass


class RequireStage(PipelineStage, ABC):
    """Represents a 'require' stage: require <variables>"""

    @abstractmethod
    def variables(self) -> Iterator["Variable"]:
        pass


class OffsetStage(PipelineStage, ABC):
    """Represents an 'offset' stage: offset <offset>"""

    @abstractmethod
    def offset(self) -> int:
        pass


class LimitStage(PipelineStage, ABC):
    """Represents a 'limit' stage: limit <limit>"""

    @abstractmethod
    def limit(self) -> int:
        pass


class DistinctStage(PipelineStage, ABC):
    """Represents a 'distinct' stage."""
    pass


class ReduceStage(PipelineStage, ABC):
    """Represents a 'reduce' stage: reduce <reducers> groupby <groupby>"""

    @abstractmethod
    def group_by(self) -> Iterator["Variable"]:
        """The variables to group by."""
        pass

    @abstractmethod
    def reduce_assignments(self) -> Iterator["ReduceStage.ReduceAssignment"]:
        """The reducer assignments."""
        pass

    class ReduceAssignment(ABC):
        """An assignment of a reducer to a variable."""

        @abstractmethod
        def assigned(self) -> "Variable":
            pass

        @abstractmethod
        def reducer(self) -> "Reducer":
            pass
